File: Sources/Hilbert/hilbert_index.cc
// Copyright 2018 The Simons Foundation, Inc. - All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "hilbert_index.hpp"

#include <cmath>

#include "Utils/next_variation.hpp"

namespace netket {

HilbertIndex::HilbertIndex(std::vector<double> localstates, int local_size,
                           int size)
    : localstates_(std::move(localstates)),
      localsize_(local_size),
      size_(size) {
  Init();
}

void HilbertIndex::Init() {
  nstates_ = std::pow(localsize_, size_);

  std::size_t ba = 1;
  for (int s = 0; s < size_; s++) {
    basis_.push_back(ba);
    ba *= localsize_;
  }

  for (std::size_t k = 0; k < localstates_.size(); k++) {
    statenumber_[localstates_[k]] = k;
  }
}

// converts a vector of quantum numbers into the unique integer identifier
std::size_t HilbertIndex::StateToNumber(const Eigen::VectorXd &v) const {
  std::size_t number = 0;

  for (int i = 0; i < size_; i++) {
    assert(statenumber_.count(v(size_ - i - 1)) > 0);
    number += statenumber_.at(v(size_ - i - 1)) * basis_[i];
  }

  return number;
}

// converts a vector of quantum numbers into the unique integer identifier
// this version assumes that a number representaiton is already known for the
// given vector v, and this function is used to update it
std::size_t HilbertIndex::DeltaStateToNumber(
    const Eigen::VectorXd &v, nonstd::span<const int> connector,
    nonstd::span<const double> newconf) const {
  std::size_t number = 0;

  for (int k = 0; k < connector.size(); k++) {
    const int ich = connector[k];
    assert(statenumber_.count(v(ich)) > 0);
    assert(statenumber_.count(newconf[k]) > 0);
    number -= statenumber_.at(v(ich)) * basis_[size_ - ich - 1];
    number += statenumber_.at(newconf[k]) * basis_[size_ - ich - 1];
  }

  return number;
}

// converts an integer into a vector of quantum numbers
Eigen::VectorXd HilbertIndex::NumberToState(int i) const {
  Eigen::VectorXd result = Eigen::VectorXd::Constant(size_, localstates_[0]);

  int ip = i;

  int k = size_ - 1;

  while (ip > 0) {
    assert(static_cast<std::size_t>(ip % localsize_) < localstates_.size());
    result(k) = localstates_[ip % localsize_];
    ip /= localsize_;
    k--;
  }
  return result;
}

constexpr int HilbertIndex::MaxStates;

}  // namespace netket

File: netket/hilbert/hilbert_index.py
# Copyright 2021 The NetKet Authors - All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
from numba.experimental import jitclass
from numba import int64, float64


spec = [
    ("_local_states", float64[:]),
    ("_local_size", int64),
    ("_size", int64),
    ("_basis", int64[:]),
]


@jitclass(spec)
class HilbertIndex:
    def __init__(self, local_states, size):
        self._local_states = np.sort(local_states)
        self._local_size = len(self._local_states)
        self._size = size

        self._basis = np.zeros(size, dtype=np.int64)
        ba = 1
        for s in range(size):
            self._basis[s] = ba
            ba *= self._local_size

    def _local_state_number(self, x):
        return np.searchsorted(self.local_states, x)

    @property
    def n_states(self):
        return self._local_size ** self._size

    @property
    def local_states(self):
        return self._local_states

    def state_to_number(self, state):
        if state.ndim != 1:
            raise RuntimeError("Invalid input shape, expecting a 1d array.")

        # Converts a vector of quantum numbers into the unique integer identifier
        number = 0

        for i in range(self._size):
            number += (
                self._local_state_number(state[self._size - i - 1]) * self._basis[i]
            )

        return number

    def number_to_state(self, number, out=None):

        if out is None:
            out = np.empty(self._size)
        # else:
        #     assert out.size == self._size

        out.fill(self._local_states[0])

        ip = number
        k = self._size - 1
        while ip > 0:
            out[k] = self._local_states[ip % self._local_size]
            ip = ip // self._local_size
            k -= 1

        return out

    def states_to_numbers(self, states, out=None):
        if states.ndim != 2:
            raise RuntimeError("Invalid input shape, expecting a 2d array.")

        if out is None:
            out = np.empty(states.shape[0], np.int64)
        # else:
        #     assert out.size == states.shape[0]

        for i in range(states.shape[0]):
            out[i] = 0
            for j in range(self._size):
                out[i] += (
                    self._local_state_number(states[i, self._size - j - 1])
                    * self._basis[j]
                )
        return out

    def numbers_to_states(self, numbers, out=None):
        if numbers.ndim != 1:
            raise RuntimeError("Invalid input shape, expecting a 1d array.")

        if out is None:
            out = np.empty((numbers.shape[0], self._size))
        # else:
        #     assert out.shape == (numbers.shape[0], self._size)

        for i in range(numbers.shape[0]):
            out[i] = self.number_to_state(numbers[i])

        return out

    def all_states(self, out=None):
        if out is None:
            out = np.empty((self.n_states, self._size))

        for i in range(self.n_states):
            self.number_to_state(i, out[i])
        return out
